src/metabase/sync_database/classify.clj

```diff
@@ -0,0 +1,204 @@
+   this used to be part of collecting field values"
+  [fingerprint field-stats]
+  (if-let [guessed-initial-type (infer-special-type/infer-field-special-type (:name fingerprint) (:base_type fingerprint))]
+    (assoc field-stats :special-type guessed-initial-type)
+    field-stats))
+
+(defn test:primary-key
+  "if a field is a primary key, it's special type must be :type/PK"
+  [fingerprint field-stats]
+  (if (:is_pk fingerprint)
+    (assoc field-stats :special-type :type/PK)
+    field-stats))
+
+(defn test:foreign-key
+  "if a field is a foreign key, it's special type must be :type/FK"
+  [fingerprint field-stats]
+  (if (:is_fk fingerprint)
+    (assoc field-stats :special-type :type/FK)
+    field-stats))
+
+(defn test:new-field
+  "Do the various tests that should only be done for a new `Field`.
+   We only run most of the field analysis work when the field is NEW in order to favor performance of the sync process."
+  [fingerprint field-stats]
+  (->> field-stats
+       (test:initial-guess      fingerprint)
+       (test:no-preview-display fingerprint)
+       (test:url-special-type   fingerprint)
+       (test:json-special-type  fingerprint)
+       (test:email-special-type fingerprint)
+       (test:category-type      fingerprint)
+       (test:primary-key        fingerprint)
+       (test:foreign-key        fingerprint)))
+
+(defn classify-table-fields
+  "Classify a table fingerprint along with all it's fields"
+  [table-fingerprint field-fingerprints]
+  {:row_count (:row_count table-fingerprint)
+   :fields (map #(test:new-field % {:id (:id %)}) field-fingerprints)})
+
+(defn classify-and-save-table!
+  "Analyze the data shape for a single `Table`."
+  [driver {table-id :id, :as table}]
+  (let [fields (table/fields table)
+        field-fingerprints (db/select FieldFingerprint :table_id table-id)
+        table-fingerprint  (db/select TableFingerprint :table_id table-id)]
+    (when-let [table-stats (u/prog1 (classify-table-fields table-fingerprint field-fingerprints)
+                             (when <>
+                               (schema/validate i/AnalyzeTable <>)))]
+      (doseq [{:keys [id preview-display special-type]} (:fields table-stats)]
+        (when (and id (or preview-display special-type))
+          (db/update-non-nil-keys! field/Field id
+            ;; if a field marked `preview-display` as false then set the visibility
+            ;; type to `:details-only` (see models.field/visibility-types)
+            :visibility_type (when (false? preview-display) :details-only)
+            :special_type    special-type
+            #_:last_analyzed   #_(u/new-sql-timestamp)))))
+
+    (db/update-where! field/Field {:table_id        table-id
+                                   :visibility_type [:not= "retired"]}
+      :last_analyzed (u/new-sql-timestamp))))
+
+(defn classify-tables!
+  "classify and save all previously saved fingerprints for tables in this database"
+  [driver {database-id :id, :as database}]
```

Comment:
nope, and another function falls out of the callgraph when it's removed. good catch!